R/serializer.R

```diff
@@ -17,30 +8,267 @@ NULL
+      serializer_identity_(val, req, res, errorHandler)
+    }, error = function(err) {
+      errorHandler(req, res, err)
     })
   }
 }
 
-.globals$serializers[["null"]] <- nullSerializer
+
+
+#' @describeIn serializers Adds a `Content-Type` header to the response object
+# name can not change!
+#' @param type The value to provide for the `Content-Type` HTTP header.
+#' @export
+serializer_content_type <- function(type, serialize_fn = identity) {
+  if (missing(type)){
+    stop("You must provide the custom content type to the serializer_content_type")
+  }
+
+  serializer_headers(
+    list("Content-Type" = type),
+    serialize_fn
+  )
+}
+
+#' @describeIn serializers CSV serializer. See \code{\link[readr:format_delim]{readr::format_csv()}} for more details.
+#' @export
+serializer_csv <- function(...) {
+  if (!requireNamespace("readr", quietly = TRUE)) {
+    stop("`readr` must be installed for `serializer_csv` to work")
+  }
+
+  serializer_content_type("text/csv; charset=UTF-8", function(val) {
+    readr::format_csv(val, ...)
+  })
+}
+
+
+
+#' @describeIn serializers HTML serializer
+#' @export
+serializer_html <- function() {
+  serializer_content_type("text/html; charset=UTF-8")
+}
+
+
+#' @describeIn serializers JSON serializer. See `jsonlite::toJSON` for more details.
+#' @export
+serializer_json <- function(...) {
+  serializer_content_type("application/json; charset=UTF-8", function(val) {
+    toJSON(val, ...)
+  })
+}
+
+#' @describeIn serializers JSON serializer with `auto_unbox` defaulting to `TRUE`. See jsonlite::toJSON` for more details.
+#' @inheritParams jsonlite::toJSON
+#' @export
+serializer_unboxed_json <- function(auto_unbox = TRUE, ...) {
+  serializer_json(auto_unbox = auto_unbox, ...)
+}
+
+
+
+
+#' @describeIn serializers RDS serializer. See [serialize()] for more details.
+#' @inheritParams base::serialize
+#' @export
+serializer_rds <- function(version = "2", ascii = FALSE, ...) {
+  if (identical(version, "3")) {
+    if (package_version(R.version) < "3.5") {
+      stop(
+        "R versions before 3.5 do not know how to serialize with `version = \"3\"`",
+        "\n Current R version: ", as.character(package_version(R.version))
+      )
+    }
+  }
+  serializer_content_type("application/octet-stream", function(val) {
+    base::serialize(val, NULL, ascii = ascii, version = version, ...)
+  })
+}
+
+#' @describeIn serializers YAML serializer. See [yaml::as.yaml()] for more details.
+#' @export
+serializer_yaml <- function(...) {
+  if (!requireNamespace("yaml", quietly = TRUE)) {
+    stop("yaml must be installed for the yaml serializer to work")
+  }
+  serializer_content_type("application/x-yaml; charset=UTF-8", function(val) {
+    yaml::as.yaml(val, ...)
+  })
+}
+
+#' @describeIn serializers Text serializer. See [format()] for more details.
+#' @export
+serializer_text <- function(...) {
+  serializer_content_type("text/plain; charset=UTF-8", function(val) {
+    format(val, ...)
+  })
+}
```

Comment:
Consider renaming to `serializer_format()` and adding  `serializer_print()` and maybe `serializer_text()` as `capture.output(cat(x))`

Reply:
(for consistency with shiny)